src/NuGet.Core/NuGet.Versioning/NuGet.Versioning.csproj

```diff
@@ -7,6 +7,7 @@
     <PackProject>true</PackProject>
     <Shipping>true</Shipping>
     <IncludeInVSIX>true</IncludeInVSIX>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
```

Comment:
The linked issue isn't clearly defined, making it sound like the intention is to enforce all style rules, not just documentation files, and also across all projects, not just one or two. The PR description also doesn't explain either why only XMLDoc, or why only 1 or 2 projects.

Most of the `src/NuGet.Core/*` projects are nupkgs published to nuget.org.  Do the other project already generate XMLDoc, and NuGet.Versioning was the only one that wasn't?

Should the linked issue be changed to explicitly mention it's only about XMLDoc, and not other style issues?

Reply:
I agree with Andy that we should clarify the scope of the PR/issue.

Reply:
Okay I updated the issue. I don't really have a reason why I picked these two projects, I just needed to start somewhere. 

One thing though, a know that <GenerateDocumentationFile> is set to true for projects with <PackProject> set to true https://github.com/NuGet/NuGet.Client/blob/15ad272450c71cdc32b8877a90ca5e6a3cb5c8e3/build/common.targets#L39-L42

Despite that I still don't get build errors and for using statements until I specify GenerateDocumentationFile in the csproj.

Reply:
can you upload a binlog for the project in question?
You can do it in teams if that's easier.

Reply:
Okay I moved this pack from the common.targets to common.project.props, and that worked. Is there a better location for it? Basically, we need to set it in a different location because it's getting set by another config.

Reply:
You said common.project.props, but I don't see that change in this PR.

Reply:
Okay that change didn't actually work but now I think I properly understand what's happening.

The <GenerateDocumentationFile> property is being set in Microsoft.NET.Sdk.BeforeCommon.targets 

```xml
  <!-- Handle XML documentation file settings -->
  <PropertyGroup Condition="'$(GenerateDocumentationFile)' == ''" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <GenerateDocumentationFile Condition="'$(DocumentationFile)' == ''">false</GenerateDocumentationFile>
    <GenerateDocumentationFile Condition="'$(DocumentationFile)' != ''">true</GenerateDocumentationFile>
  </PropertyGroup>
  <PropertyGroup Condition="'$(GenerateDocumentationFile)' == 'true' and '$(DocumentationFile)' == ''" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <DocumentationFile Condition="'$(MSBuildProjectExtension)' == '.vbproj'">$(AssemblyName).xml</DocumentationFile>
    <DocumentationFile Condition="'$(MSBuildProjectExtension)' != '.vbproj'">$(IntermediateOutputPath)$(AssemblyName).xml</DocumentationFile>
  </PropertyGroup>
  <PropertyGroup Condition="'$(GenerateDocumentationFile)' != 'true'" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <DocumentationFile />
  </PropertyGroup>
```

Looking at the msbuild generated properties file the next file up the tree we control is the individual csproj, which is where we set the <PackProject> property.

Moving the code to the common.project.props file doesn't work because when it goes to evaluate the condition <PackProject> is empty, so <GenerateDocumentationFile> is set to false in Microsoft.NET.Sdk.BeforeCommon.targets. 

Even if we set the <GenerateDocumentationFile> value in the csproj the value for<DocumentationFile> will be set by Microsoft.NET.Sdk.BeforeCommon.targets instead of what we have in common.targets
```xml
<PropertyGroup Condition=" '$(PackProject)' == 'true' ">
    <GenerateDocumentationFile Condition=" '$(GenerateDocumentationFile)' == '' ">true</GenerateDocumentationFile>
    <DocumentationFile Condition=" '$(DocumentationFile)' == '' AND '$(GenerateDocumentationFile)' == 'true' AND '$(IsNetCoreProject)' != 'true' ">$(OutputPath)\$(AssemblyName).xml</DocumentationFile>
</PropertyGroup>
```

Is there a reason to put the xml in the intermediate output path instead of the output path? If not then I think we could delete the block above from common.targets and set <GenerateDocumentationFile> for all projects we are also packing and currently don't have it set to false.

Reply:
> Moving the code to the common.project.props file doesn't work because when it goes to evaluate the condition is empty, so is set to false in Microsoft.NET.Sdk.BeforeCommon.targets.

Which condition is false?

Reply:
> Is there a reason to put the xml in the intermediate output path instead of the output path? If not then I think we could delete the block above from common.targets and set for all projects we are also packing and currently don't have it set to false.

No clue. I'm guessing it's missing defaults based on the is net core check.

Reply:
Keep in mind that the NuGet.Client repo has existed for years, before the .NET (Core) SDK was a thing, so while the SDK might now have defaults to make it work out of the box, perhaps in the past it didn't, or pre-SDK style projects might have worked differently as well.

I prefer a common config to apply to all projects, rather than opting individual projects into docs.

Reply:
> > Moving the code to the common.project.props file doesn't work because when it goes to evaluate the condition is empty, so is set to false in Microsoft.NET.Sdk.BeforeCommon.targets.
> 
> Which condition is false?
This one, since PackProject is set later in the file it has no value when this is evaluated. 
```xml
<PropertyGroup Condition=" '$(PackProject)' == 'true' ">
```

Reply:
> I prefer a common config to apply to all projects, rather than opting individual projects into docs.

Agreed but where that common config should live is difficult because of how `<GenerateDocumentationFile>` is set by `Microsoft.NET.Sdk.BeforeCommon.targets`. 

Since we now know that without `<GenerateDocumentationFile>` we can't enforce all the style rules, is it worth just enabling it for everyone?

Reply:
I think it's ok to enable it for everyone. 
We'd probably need to disable the IDE0005 in some projects where it fails,.
I can't think of anything else.

Reply:
@zivkan to focus this conversation here because this has the most context. So, the latest change I pushed is what it looks like if we enable `GenerateDocumentationFile` for all projects all at once. I kept flipping back and forth between enabling it per project then once all projects have had it enabled moving it to the central build props project. 

Currently we do not generate the documentation file for any of our projects because of the way the value is determined. So, I believe options for moving forward are:
1. Enable `GenerateDocumentationFile` for all projects, using `NoWarn` to suppress build errors, then overtime remove the suppressions per project and fix the errors that come up. (current state)
2.  Enable `GenerateDocumentationFile` for all projects, using `WarningsNotAsErrors` to suppress build errors, then overtime remove the suppressions per project and fix the errors that come up.
3. Enable `GenerateDocumentationFile` for individual projects, fixing the build errors as it is enabled. Once all projects have been updated to generate documentation files, we can move it to a central location. 
4. Force `GenerateDocumentationFile` to true for projects which have `PackProject` set to true. This will limit the scope of how many projects need to be fixed upfront, however it will still be a lot. We would need to selectively turn on `GenerateDocumentationFile` for all projects in order to get the style fixes like IDE0005 to show up as errors in the build.